### PR TITLE
TS-4622 Use port from SRV response when connecting to origin

### PR DESCRIPTION
**Pull request by jacksontj**

### CODE REVIEW

**Comment by jpeach** [comment]
-1, this breaks ``TSHttpTxnServerAddrSet``.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/388/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/284/ for details.

**Comment by jpeach** [comment]
@jacksontj Maybe the right place to do this is when ``HttpSM::set_next_state()
`` handles ``SM_ACTION_DNS_LOOKUP``?

**Comment by jacksontj** [comment]
Yea, I'll have to look a bit more. Ideally its as close to that DNS lookup as possible, that way if some plugin wants to overwrite it I've done the change in core before they can (so that we won't be clobbering eachother). Have to look into it this week.

**Comment by jacksontj** [comment]
@jpeach `HttpTransact::OSDNSLookup` seems to be the correct place-- as it is called immediately after the DNS lookup is complete-- and is what sets the port numbers for the destination.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/455/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/349/ for details.

File: proxy/http/HttpTransact.cc [hunk 1]
```diff
@@ -1783,7 +1783,13 @@ HttpTransact::OSDNSLookup(State *s)
...
   // been provided.
   ats_ip_copy(&s->server_info.dst_addr, s->host_db_info.ip());
-  s->server_info.dst_addr.port() = htons(s->hdr_info.client_request.port_get()); // now we can set the port.
+  // TODO ideally only if ports aren't defined in remap
+  // If the SRV response has a port number, we should honor it. Otherwise we do the port defined in remap
+  if (s->dns_info.srv_port) {
+    s->server_info.dst_addr.port() = htons(s->dns_info.srv_port);
+  } else {
+    s->server_info.dst_addr.port() = htons(s->hdr_info.client_request.port_get()); // now we can set the port.
+  }
```
Comment (JamesPeach) [inline proxy/http/HttpTransact.cc]: If a plugin used ``TSHttpTxnServerAddrSet()``, then at this point ``dst_addr`` may be the address set by the plugin in which case we should not clobber it. However it is already being clobbered from the request URL, so maybe the cases I tested were already broken (I was using port 80 and 443).

Setting the port from the SRV record should probably be conditional on ``srv_lookup_success``.

Comment (jacksontj) [inline proxy/http/HttpTransact.cc]: I'll change the conditional-- that definitely makes sense.

From looking at the API callout code-- it does seem like its already broken :/

It seems that `TSHttpTxnServerAddrSet()` is called before the DNS lookup-- and if so it bypasses the [dns lookup| https://github.com/apache/trafficserver/blob/master/proxy/http/HttpSM.cc#L7179]. So, I would think in this section of HttpTransact we would want to skip all port setting if `api_server_addr_set` is true.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/458/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/354/ for details.

**Comment by jacksontj** [comment]
@jpeach 

I did some testing of my own and I can't reproduce any problems with `TsHttpTxnServerAddrSet()` at all. TSHttpTxnServerAddrSet() is supposed to be called before the DNS lookup. If called then the core bypasses the entire OS_DNS stuff which includes this port setting block in HttpTransact.

I tested calling the API on both the `TS_EVENT_HTTP_READ_REQUEST_HDR` and `TS_EVENT_HTTP_OS_DNS` hooks-- with no problem at all.

You mentioned in a previous comment that it was broken in your testing, can you share the code you were using to see it broken?

**Comment by jpeach** [comment]
👍 This looks good to me and I think it will be safe w/ "TsHttpTxnServerAddrSet()".

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/464/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/360/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/465/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/466/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/361/ for details.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/362/ for details.

**Pull request merged**